test/e2e/helpers_gateway_test.go

```diff
@@ -0,0 +1,296 @@
+				Port:     gatewayv1alpha2.PortNumber(80),
+			}},
+		},
+	}
+	gw, err = gc.GatewayV1alpha2().Gateways(corev1.NamespaceDefault).Create(ctx, gw, metav1.CreateOptions{})
+	require.NoError(t, err)
+	return gw
+}
+
+func verifyGateway(ctx context.Context, t *testing.T, env environments.Environment, gw *gatewayv1alpha2.Gateway) {
+	gc, err := gatewayclient.NewForConfig(env.Cluster().Config())
+	require.NoError(t, err)
+
+	t.Log("verifying that the gateway receives a final ready condition once reconciliation completes")
+	require.Eventually(t, func() bool {
+		gw, err = gc.GatewayV1alpha2().Gateways(corev1.NamespaceDefault).Get(ctx, gw.Name, metav1.GetOptions{})
+		require.NoError(t, err)
+		for _, cond := range gw.Status.Conditions {
+			if cond.Reason == string(gatewayv1alpha2.GatewayReasonReady) {
+				return true
+			}
+		}
+		return false
+	}, gatewayUpdateWaitTime, time.Second)
+}
+
+func deployGatewayWithTCPListener(ctx context.Context, t *testing.T, env environments.Environment) *gatewayv1alpha2.Gateway {
+	gc, err := gatewayclient.NewForConfig(env.Cluster().Config())
+	require.NoError(t, err)
+
+	t.Log("deploying a supported gatewayclass to the test cluster")
+	supportedGatewayClass := &gatewayv1alpha2.GatewayClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: uuid.NewString(),
+		},
+		Spec: gatewayv1alpha2.GatewayClassSpec{
+			ControllerName: gateway.ControllerName,
+		},
+	}
+	supportedGatewayClass, err = gc.GatewayV1alpha2().GatewayClasses().Create(ctx, supportedGatewayClass, metav1.CreateOptions{})
+	require.NoError(t, err)
+
+	t.Log("deploying a gateway to the test cluster using unmanaged gateway mode")
+	gw := &gatewayv1alpha2.Gateway{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "kong",
+		},
+		Spec: gatewayv1alpha2.GatewaySpec{
+			GatewayClassName: gatewayv1alpha2.ObjectName(supportedGatewayClass.Name),
+			Listeners: []gatewayv1alpha2.Listener{
+				{
+					Name:     "http",
+					Protocol: gatewayv1alpha2.HTTPProtocolType,
+					Port:     gatewayv1alpha2.PortNumber(80),
+				},
+				{
+					Name:     "tcp",
+					Protocol: gatewayv1alpha2.TCPProtocolType,
+					Port:     gatewayv1alpha2.PortNumber(tcpListnerPort),
+				},
+			},
+		},
+	}
+	_, err = gc.GatewayV1alpha2().Gateways(corev1.NamespaceDefault).Get(ctx, gw.Name, metav1.GetOptions{})
+	if err == nil {
+		t.Logf("gateway %s exists, delete and re-create it", gw.Name)
+		err = gc.GatewayV1alpha2().Gateways(corev1.NamespaceDefault).Delete(ctx, gw.Name, metav1.DeleteOptions{})
+		require.NoError(t, err)
+		gw, err = gc.GatewayV1alpha2().Gateways(corev1.NamespaceDefault).Create(ctx, gw, metav1.CreateOptions{})
+		require.NoError(t, err)
+	} else {
+		require.True(t, kerrors.IsNotFound(err))
+		gw, err = gc.GatewayV1alpha2().Gateways(corev1.NamespaceDefault).Create(ctx, gw, metav1.CreateOptions{})
+		require.NoError(t, err)
+	}
+	return gw
+}
+
+func deployHTTPRoute(ctx context.Context, t *testing.T, env environments.Environment, gw *gatewayv1alpha2.Gateway) {
```

Comment:
I  wonder how one could access this HTTPRoute given that it's not returned from this `func`. There are no labels either so basically there's no way to query it if I get this right. Wouldn't that be a problem where this `func` is called?

Reply:
As far as I can see now, this has only moved in this PR so I'm not going to push this by any means but it might be worth taking a look at.

Reply:
It defined a path match in its spec, so we could use `http://gatewayIP/path` to access it.

Reply:
That would be nice to put in a comment

Reply:
added comments.